cfgrammar/src/lib/header.rs

```diff
@@ -335,6 +395,110 @@ impl<'input> GrmtoolsSectionParser<'input> {
     }
 }
 
+#[derive(Debug, Eq, PartialEq)]
+pub struct Header {
+    contents: HashMap<String, (Span, Value)>,
+    used: Vec<String>,
+}
+
+impl Header {
+    pub fn new() -> Self {
+        Header {
+            contents: HashMap::new(),
+            used: vec![],
+        }
+    }
+
+    pub fn contents(&self) -> &HashMap<String, (Span, Value)> {
+        &self.contents
+    }
+
+    pub fn contents_mut(&mut self) -> &mut HashMap<String, (Span, Value)> {
+        &mut self.contents
+    }
+
+    pub fn mark_key_used(self: &mut Self, key: String) {
+        let pos = self.used.binary_search(&key);
+        match pos {
+            // Already Used
+            Ok(_) => {}
+            Err(pos) => {
+                self.used.insert(pos, key);
+            }
+        }
+    }
+
+    pub fn unused(&self) -> impl Iterator<Item = (&String, &(Span, Value))> {
+        // On the happy path this is empty. It might be faster to add the map to a vec.
+        // then call `sort_by_key` from the first element to compare against the sorted
+        // `used` field in a single pass. However that is probably overkill and this is
+        // simple to implement.
+        self.contents
+            .iter()
+            .filter(|(key, _)| self.used.binary_search(key).is_ok())
+    }
+
+    pub fn query_key(
+        &self,
+        key: &str,
+        query_mask: u16,
```

Comment:
Of course there is some borrow checker issues with the combination of `query_key`/`mark_key_unused`,
So this interface seems to require some changes.  I haven't gotten the entry API to work (since we need to mutate the `header` not the `HashMap`. So might have to resort to `contains_key` -> `mutate` -> `query`.

Reply:
Should be fixed in [06458b1](https://github.com/softdevteam/grmtools/pull/547/commits/06458b1fcd1ac37d837edaba89d2ef8e5db919ff)

I'm not sure it's the best fix, the "mark as used" even on error behavior
is intended to avoid the following situation:

1. specify `%grmtools{yacckind: Foo}` 
2.  Get an "invalid yacc kind" for `Foo` error.
3.  *also* get `unused yacckind`